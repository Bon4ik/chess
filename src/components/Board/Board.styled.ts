import styled from "styled-components";

import { COLOR_PALETTE } from "utils/const";

export const BoardContainer = styled.div`
  width: 800px;
  height: 800px;
  border: 8px solid ${COLOR_PALETTE.BOARD_DARK};
  flex-wrap: wrap;
  display: flex;
`;
