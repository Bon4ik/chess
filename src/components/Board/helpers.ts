import { CHESS_PIECE_TYPE, CHESS_PIECES } from "utils/const";

export const isPawnMoveValid = (
  [targetRow, targetCol]: number[],
  [selectedRow, selectedCol]: number[],
  boardMatrix: IBoardCellEntity[][],
  activePlayer: string
): boolean => {
  // return false if it's the same cell
  if (selectedRow === targetRow && selectedCol === targetCol) return false;

  // if it's attack
  if (boardMatrix[targetRow][targetCol].value) {
    const pawnAttackCells = [
      [-1, -1],
      [-1, 1],
      [1, 1],
      [1, -1],
    ];

    return pawnAttackCells.some(
      ([row, col]) =>
        row + selectedRow === targetRow && col + selectedCol === targetCol
    );
  }

  // if it's move
  if (targetCol !== selectedCol) return false;
  else if (activePlayer === CHESS_PIECE_TYPE.DARK) {
    const diffRow = targetRow - selectedRow;

    if (diffRow === 1) return true;
    else if (diffRow === 2 && selectedRow === 1) return true;
    else return false;
  } else {
    const diffRow = selectedRow - targetRow;

    if (diffRow === 1) return true;
    else if (diffRow === 2 && selectedRow === 6) return true;
    else return false;
  }
};

export const isMoveValid = (
  targetCoordinates: number[],
  selecteCoordinates: number[],
  boardMatrix: IBoardCellEntity[][],
  activePlayer: string
): boolean => {
  const [selectedRow, selectedCol] = selecteCoordinates;
  const chessPiece = boardMatrix[selectedRow][selectedCol].value?.toLowerCase();

  switch (chessPiece) {
    case CHESS_PIECES.PAWN:
      return isPawnMoveValid(
        targetCoordinates,
        selecteCoordinates,
        boardMatrix,
        activePlayer
      );
    default:
      return false;
  }
};
