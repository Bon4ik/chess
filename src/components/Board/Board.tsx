import React, { useMemo, useCallback, Fragment, useState } from "react";

import BoardCell from "./BoardCell";
import BottomBar from "./BottomBar";
import { BoardContainer } from "./Board.styled";
import { isMoveValid } from "./helpers";
import {
  CHESS_PIECES,
  FREE_LEGAL_SLOTS,
  DEFAULT_PAWN_ROW,
  CHESS_PIECE_TYPE,
  DEFAULT_BOARD,
} from "utils/const";
import { getDefaultMatrix } from "utils/getDefaultBoardMatrix";

const BoardComponent = () => {
  const [freeLegalSlots, setFreeLegalSlots] = useState(FREE_LEGAL_SLOTS);
  const [boardMatrix, setBoardMatrix] = useState(getDefaultMatrix(DEFAULT_BOARD));
  const [activePlayer, setActivePlayer] = useState(CHESS_PIECE_TYPE.LIGHT);
  const [selectedChessPiece, setSelectedChessPiece] = useState<number[]>([]);

  const onCellClick = useCallback(
    (targetCoordinates: number[], chessPieceType?: string) => {
      if (chessPieceType === activePlayer) {
        setSelectedChessPiece(targetCoordinates);
      } else if (selectedChessPiece.length) {
        const isValid = isMoveValid(
          targetCoordinates,
          selectedChessPiece,
          boardMatrix,
          activePlayer
        );

        if (isValid) {
          const nextActivePlayer =
            activePlayer === CHESS_PIECE_TYPE.LIGHT
              ? CHESS_PIECE_TYPE.DARK
              : CHESS_PIECE_TYPE.LIGHT;
          const newBoardMatrix = boardMatrix.map((row) => [...row]);
          const [newRow, newCol] = targetCoordinates;
          const [oldRow, oldCol] = selectedChessPiece;

          // update new cell
          newBoardMatrix[newRow][newCol].value =
            boardMatrix[oldRow][oldCol].value;
          newBoardMatrix[newRow][newCol].chessPieceType = activePlayer;

          // remove old cell
          newBoardMatrix[oldRow][oldCol].value = "";

          setSelectedChessPiece([]);
          setBoardMatrix(newBoardMatrix);
          setActivePlayer(nextActivePlayer);
        }
      }
    },
    [activePlayer, boardMatrix, selectedChessPiece]
  );

  const addWhitePawn = useCallback(() => {
    if (freeLegalSlots.length) {
      const pawnIndex = Math.floor(Math.random() * freeLegalSlots.length);
      const newFreeLegalSlots = freeLegalSlots
        .slice(0, pawnIndex)
        .concat(freeLegalSlots.slice(pawnIndex + 1));

      const newBoardMatrix = boardMatrix.map((row) => [...row]);
      newBoardMatrix[DEFAULT_PAWN_ROW][freeLegalSlots[pawnIndex]] = {
        ...newBoardMatrix[DEFAULT_PAWN_ROW][freeLegalSlots[pawnIndex]],
        value: CHESS_PIECES.PAWN,
        chessPieceType: CHESS_PIECE_TYPE.LIGHT,
      };

      setBoardMatrix(newBoardMatrix);
      setFreeLegalSlots(newFreeLegalSlots);
    }
  }, [boardMatrix, freeLegalSlots]);

  const board = useMemo(
    () =>
      boardMatrix.map((row: IBoardCellEntity[]) =>
        row.map((cell: IBoardCellEntity) => (
          <BoardCell
            {...cell}
            key={cell.id}
            onCellClick={onCellClick}
            isActive={
              selectedChessPiece[0] === cell.coordinates[0] &&
              selectedChessPiece[1] === cell.coordinates[1]
            }
          />
        ))
      ),
    [boardMatrix, onCellClick, selectedChessPiece]
  );

  const isAddPawnDisabled = useMemo(() => freeLegalSlots.length === 0, [
    freeLegalSlots,
  ]);

  return (
    <Fragment>
      <BoardContainer>{board}</BoardContainer>
      <BottomBar addPawn={addWhitePawn} isAddPawnDisabled={isAddPawnDisabled} />
    </Fragment>
  );
};

export default BoardComponent;
