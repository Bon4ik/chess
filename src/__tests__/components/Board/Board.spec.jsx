import React from "react";
import renderer from "react-test-renderer";

import Board from "components/Board";

describe("BottomBar", () => {
  it("should renders correctly", () => {
    const tree = renderer.create(<Board />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
