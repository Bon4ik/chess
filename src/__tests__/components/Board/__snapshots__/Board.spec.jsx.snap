// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomBar should renders correctly 1`] = `
Array [
  <div
    className="sc-eCApGN wtGhL"
  >
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu gDJAzW"
        type="DARK"
      >
        p
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx bsOIpA"
      onClick={[Function]}
      type="DARK"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
    <div
      className="sc-bdnylx kllbOP"
      onClick={[Function]}
      type="LIGHT"
    >
      <span
        className="sc-gtssRu hTGNSe"
      >
        
      </span>
    </div>
  </div>,
  <div
    className="sc-hKFyIo hdzxak"
  >
    <button
      className="sc-dlnjPT ehpGlk"
      disabled={false}
      onClick={[Function]}
    >
      Add pawn
    </button>
  </div>,
]
`;
